torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -324,127 +258,141 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
         self._broadcast_params()
 
         # Sync hypothethical new results from the wrapped optimizer to the exposed param_groups
-        self._update_param_groups(local_to_global=True)
+        ZeroRedundancyOptimizer._sync_param_groups(self.optim.param_groups, self.param_groups)
 
         return loss
 
-    def load_local_state_dict(self, state_dict: dict) -> None:
-        """Loads this rank's state_dict.
-
-        .. warning: This is not meant to load the global state dict.
+    def state_dict(self) -> Dict[str, Any]:
```

Comment:
so, this is a big change: when the state_dict is queried, rebuild a "normal" dict out of all the shards. There's some indexing trickery (since all the shards live in their own sub parameter space, which is a partition randomly selected more or less), but the end result is that the state is the one from a non-sharded optimizer:
- compatible load/save with a "normal" pytorch job
- a normal pytorch job can load these dicts